src/pwr.rs

```diff
@@ -1,13 +1,145 @@
+}
 
 pub struct Pwr {
     pub cr1: CR1,
     pub cr2: CR2,
     pub cr3: CR3,
     pub cr4: CR4,
+    pub scr: SCR,
+    pub sr1: SR1,
+}
+
+impl Pwr {
+    /// Configures dynamic voltage regulator range
+    ///
+    /// Will panic if low-power range is selected for higher system clock
+    pub fn set_power_range(&mut self, range: VosRange, clocks: &Clocks) -> Result<(), Error> {
+        match range {
+            VosRange::HighPerformance => unsafe {
+                {
+                    self.cr1
+                        .reg()
+                        .modify(|_, w| w.vos().bits(VosRange::HighPerformance as u8))
+                }
+                Ok(())
+            },
+            VosRange::LowPower => {
+                if clocks.sysclk() > 26.mhz().into() {
+                    Err(Error::SysClkTooHighVos)
+                } else {
+                    unsafe {
+                        self.cr1
+                            .reg()
+                            .modify(|_, w| w.vos().bits(VosRange::LowPower as u8))
+                    }
+                    Ok(())
+                }
+            }
+        }
+    }
+
+    /// Switches the system into low power run mode
+    pub fn low_power_run(&mut self, clocks: &Clocks) -> Result<(), Error> {
+        if clocks.sysclk() > 2.mhz().into() {
+            Err(Error::SysClkTooHighLpr)
+        } else {
+            self.cr1.reg().modify(|_, w| w.lpr().set_bit());
+            Ok(())
+        }
+    }
+
+    /// Enters 'Shutdown' low power mode.
+    pub fn shutdown(&mut self, wkup: &[WakeUpSource], scb: &mut SCB) -> ! {
+        for source in wkup {
+            match source {
+                WakeUpSource::Internal => self.cr3.reg().modify(|_, w| w.ewf().set_bit()),
+                WakeUpSource::WKUP1 => self.cr3.reg().modify(|_, w| w.ewup1().set_bit()),
+                WakeUpSource::WKUP2 => self.cr3.reg().modify(|_, w| w.ewup2().set_bit()),
+                WakeUpSource::WKUP3 => self.cr3.reg().modify(|_, w| w.ewup3().set_bit()),
+                WakeUpSource::WKUP4 => self.cr3.reg().modify(|_, w| w.ewup4().set_bit()),
+                WakeUpSource::WKUP5 => self.cr3.reg().modify(|_, w| w.ewup5().set_bit()),
+            }
+        }
+        scb.set_sleepdeep();
+        self.scr.reg().write(|w| {
+            w.wuf1()
+                .set_bit()
+                .wuf2()
+                .set_bit()
+                .wuf3()
+                .set_bit()
+                .wuf4()
+                .set_bit()
+                .wuf5()
+                .set_bit()
+                .sbf()
+                .set_bit()
+        });
+        unsafe { self.cr1.reg().modify(|_, w| w.lpms().bits(0b111)) };
+        cortex_m::asm::dsb();
+        cortex_m::asm::wfi();
+        loop {}
+    }
+
+    /// Returns the reason, why wakeup from shutdown happened. In case there is more then one,
+    /// a single random reason will be returned
+    pub fn read_wakeup_reason(&mut self) -> Option<WakeUpSource> {
```

Comment:
Would it make sense to use a `bitfield!{ .. }` here instead? That way all reasons would get forwarded to the user.

Reply:
Well, not sure about it, With `WakeUpSource` as enum we use same type to _configure_ wake up source and to _query_ wake up source. With bit field semantics will be not that obvious: in case we decide to fully replace enum with `bitfield!{}` configuration function will be even uglier than it is it right now. But even worse, `read_wakeup_reason` will return a structure with settable fields, which will be confusing. 
Another one option  will be keeping enum to configure wake up source and a bitfield to read wake up sources. But this will be too confusing and non-obvious.

What if we will return a fixed size array of `Option<WakeUpSource>` from the `read_wakeup_reason`? @korken89

Reply:
Hmm not sure I'm following, you could use the same bitfield in `shutdown` as well, just that it won't take a slice anymore.

Reply:
After re-thinking tody i decided to agree, let's go with `bitfield!` option. Will update the PR during christmas weekend